lib/widgets/action_sheet.dart

```diff
@@ -12,51 +15,201 @@ import 'actions.dart';
+    // method; see its doc). But currently it will be constant through the life of
+    // any message list, so that's fine.
+    final messageListPage = MessageListPage.ancestorOf(widget.messageListContext);
+    final isComposeBoxOffered = messageListPage.composeBoxController != null;
+    final narrow = messageListPage.narrow;
+    final isMessageRead = widget.message.flags.contains(MessageFlag.read);
+    final markAsUnreadSupported = store.connection.zulipFeatureLevel! >= 155; // TODO(server-6)
+    final showMarkAsUnreadButton = markAsUnreadSupported && isMessageRead;
+
+    final hasThumbsUpReactionVote = widget.message.reactions
+      ?.aggregated.any((reactionWithVotes) =>
+        reactionWithVotes.reactionType == ReactionType.unicodeEmoji
+        && reactionWithVotes.emojiCode == '1f44d'
+        && reactionWithVotes.userIds.contains(store.selfUserId))
+      ?? false;
+
+    final optionButtons = [
+      if (!hasThumbsUpReactionVote)
+        AddThumbsUpButton(message: widget.message, messageListContext: widget.messageListContext),
+      StarButton(message: widget.message, messageListContext: widget.messageListContext),
+      if (isComposeBoxOffered)
+        QuoteAndReplyButton(message: widget.message, messageListContext: widget.messageListContext),
+      if (showMarkAsUnreadButton)
+        MarkAsUnreadButton(message: widget.message, messageListContext: widget.messageListContext, narrow: narrow),
+      CopyMessageTextButton(message: widget.message, messageListContext: widget.messageListContext),
+      CopyMessageLinkButton(message: widget.message, messageListContext: widget.messageListContext),
+      ShareButton(message: widget.message, messageListContext: widget.messageListContext),
+    ];
+
+    // Pad the bottom inset. The left/top/right insets are already handled by
+    // `showModalBottomSheet.useSafeArea: true` above, which keeps the sheet
+    // out of those insets.
+    return SafeArea(
+      minimum: const EdgeInsets.only(bottom: 16),
+      child: Padding(
+        padding: const EdgeInsets.fromLTRB(16, 0, 16, 0),
+        child: Column(
+          crossAxisAlignment: CrossAxisAlignment.stretch,
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            // TODO(#217): show message text
+            Flexible(
+              child: Stack(
+                children: [
+                  Column(
+                    mainAxisSize: MainAxisSize.min,
+                    children: [
+                      // Serves as the top dynamic padding, which changes when
+                      // the action sheet is scrolled in the upper direction.
+                      _ScrollControllerBuilder(
+                        scrollController: scrollController,
+                        builder: (_, scrollController) => SizedBox(
+                          height: math.max(
+                            16 - scrollController.position.extentBefore,
```

Comment:
Bump on https://github.com/zulip/zulip-flutter/pull/853#discussion_r1725975882 (using 8 here instead of 16).